### DISCUSSION TITLE
Custom args with CSF3 and Typescript

### DISCUSSION THREAD

**Discussion by RomainEndelin** [storybookjs/storybook#21587] [Help]
Hi,
I'm really enjoying the CSF3 & Typescript integration, thanks for all the efforts!

But Typescript doesn't seem to accept custom args. This example was inspired by https://storybook.js.org/docs/7.0/react/writing-stories/stories-for-multiple-components#reusing-subcomponent-stories. It renders correctly in Storybook, but raises type errors:

```tsx
import type { Meta, StoryObj } from '@storybook/react'
import { FunctionComponent, PropsWithChildren } from 'react'

const List: FunctionComponent<PropsWithChildren> = ({ children }) => <ul>{children}</ul>
const ListItem: FunctionComponent<{ name: string }> = ({ name }) => <li>{name}</li>

const meta = {
  title: 'List',
  component: List,
} satisfies Meta<typeof List>

export default meta
type Story = StoryObj<typeof meta>

export const ListStory: Story = {
  render: ({ itemNames }) => (
    <List>
      {itemNames.map((itemName) => (
        <ListItem key={itemName} name={itemName} />
      ))}
    </List>
  ),
  args: {
    itemNames: ['test'],
  },
}
```

**Errors:**
```
Test.stories.tsx:16:14 - error TS2339: Property 'itemNames' does not exist on type '{ children?: ReactNode; }'.

16   render: ({ itemNames }) => (
                ~~~~~~~~~

Test.stories.tsx:24:5 - error TS2322: Type '{ itemNames: string[]; }' is not assignable to type 'Partial<{ children?: ReactNode; }> & { children?: ReactNode; }'.
  Object literal may only specify known properties, and 'itemNames' does not exist in type 'Partial<{ children?: ReactNode; }> & { children?: ReactNode; }'.

24     itemNames: ['test'],
       ~~~~~~~~~~~~~~~~~~~

node_modules/@storybook/types/dist/index.d.ts:1148:5
    1148     args?: Partial<TArgs>;
             ~~~~
    The expected type comes from property 'args' which is declared here on type 'StoryAnnotations<ReactRenderer, { children?: ReactNode; }, { children?: ReactNode; }>'


Found 2 errors in the same file, starting at: Test.stories.tsx:16
```

**Render:**
<img width="931" alt="Screenshot 2023-03-14 at 11 43 38" src="https://user-images.githubusercontent.com/1875908/224863395-8e16f0bc-4bf9-45ee-8847-e44b9d3250f6.png">

### Workaround

In this example, using the `children` args would work, but it's less friendly to control than a plain array of `string`.

### Suggestion

Would it make sense to allow an optional generic for custom args? `const meta = { ... } satisfies Meta<typeof List, { itemNames: string[] }>`

This could also be defined at the Story level:
```ts
const meta = { ... } satisfies Meta<typeof List>

type Story = StoryObj<typeof meta, { itemNames: string[] }>
```

**Comment by vyivel** [comment]
Something like this seems to type-check properly for me (Vue):

```ts
export interface ComponentProps {
  prop: number;
}

defineProps<ComponentProps>();
```

```ts
export default {
  component: Component,
} satisfies Meta<ComponentProps>;

type Story = StoryObj<ComponentProps & { storyProp: number }>;

export const Default = {
  args: {
    prop: 42,
    storyProp: 1337,
  },
} satisfies Story;
```

Basically, try using the props type directly instead of `typeof`.

**Reply by gforro** [reply]
Thanks Vyivel for the answer. This works also for React. The official documentation https://storybook.js.org/docs/react/essentials/controls#fully-custom-args  and https://storybook.js.org/docs/react/writing-stories/stories-for-multiple-components#creating-a-template-component is a bit misleading. 
The props type should be used instead of the component type (`typeof Component`), otherwise TypeScript does not add the custom arg to the render and args part of the story:
use 
`type Story = StoryObj<ComponentProps>`
instead of the everywhere suggested 
`type Story = StoryObj<typeof Component>` or `type Story = StoryObj<typeof meta>`
and You can extend the component's props with a custom arg:
`type StoryWithCustomArg = StoryObj<ComponentProps & { myCustomArg?: string }>`

with the `typeof ...` way of coding I was not able to extend the default args, which is derived from the component type.
In case, that You prefer to not export the `Props` type, You can use:
```
type ComponentPros = Parameters<typeof FuncComponent>[0];
type StoryWithCustomArg = StoryObj<ComponentProps & { myCustomArg?: string }>
```

However such definitions "drops" the advanced type checking provided through the `satisfies` - described in https://storybook.js.org/blog/improved-type-safety-in-storybook-7#pair-it-with-satisfies-for-better-type-safety
Problem is, that it is not possible to use the `StoryObj<typeof meta>`. At least I was not able to find a way of "injecting custom arg type to the default args based on the component props".

**Reply by gforro** [reply]
A possible solution, which allows to use the `typeof meta`, that is it keeps the more type safe args in the story is as it follows:
let's have a simple `FullName` component:
```jsx
function FullName({firstName, lastName}: {firstName: string, lastName: string}) {
    return <span>{firstName} {lastName}</span>;
}
```
We can use a custom arg for this component:
```jsx
import { Meta, StoryObj } from "@storybook/react";
import { FullName } from "./FullName";

type FullNameProps = Parameters<typeof FullName>[0];

const meta = {
  title: "Full Name",
  component: FullName,
  render: ({ times, ...args }: FullNameProps & { times?: number }) => {
    if (!times) {
      return <FullName {...args} />;
    }

    return (
      <div style={{ display: "flex", flexDirection: "column" }}>
        {new Array(times).fill(null).map((_, i) => (
          <FullName key={i} {...args} />
        ))}
      </div>
    );
  },
} satisfies Meta<typeof FullName>;

type Story = StoryObj<typeof meta>;

export default meta;

export const One: Story = {
  args: {
    firstName: "Happy",
    lastName: "John",
  },
};

export const Many: Story = {
  ...One,
  args: {
    ...One.args,
    times: 5,
  },
};
```

Now it is type safe, as leaving out the required args `firstName` or `lastName` cause a compilation error in TypeScript.
